components/appmgt/org.wso2.carbon.appmgt.custom.mediators/src/main/java/org/wso2/carbon/appmgt/custom/mediators/redirection/RefererRewriteMediator.java

```diff
@@ -40,32 +37,31 @@
      */
     public boolean mediate(MessageContext messageContext) {
 
-        Cache cache = this.getAppContextVersionConfigCache();
-        Map<String,String> contextVersionMap = (HashMap<String, String>) cache.get
-                (AppMConstants.APP_CONTEXT_VERSION_CACHE_KEY);
-        //ToDo: handle if cache has expired
-
-        org.apache.axis2.context.MessageContext axis2MC =
-                ((Axis2MessageContext) messageContext). getAxis2MessageContext();
-        Map map = (Map)axis2MC.getProperty(AppMConstants.TRANSPORT_HEADERS);
-                String host = (String) map.get(AppMConstants.HOST);
-        String referer = (String)map.get(AppMConstants.REFERER);
+        Map<String, String> contextVersionMap = AppContextCacheUtil.getTenantContextVersionUrlMap();
+        org.apache.axis2.context.MessageContext axis2MC = ((Axis2MessageContext) messageContext)
+                .getAxis2MessageContext();
+        Map map = (Map) axis2MC.getProperty(AppMConstants.TRANSPORT_HEADERS);
+        String host = (String) map.get(AppMConstants.HOST);
+        String referer = (String) map.get(AppMConstants.REFERER);
 
         //Identify http/https
         String transport = referer.split(AppMConstants.URL_DELIMITER)[0];
         String endpoint;
         String restFullRequestPath = (String) messageContext.getProperty("REST_FULL_REQUEST_PATH");
-        for (String contextVersion : contextVersionMap.keySet()){
-            if (referer.contains(contextVersion)){
+        for (String contextVersion : contextVersionMap.keySet()) {
+            //TODO: This iteration is too slow. we need a better data structure for this.
```

Comment:
There cannot be "TODO"s like this, this must be figured out and fixed .. A possible solution, if the version part from the "referer" can be parsed, if it's there in a well known format, that version can be used to do a key exists in the map. Then, this full operation will be executing as a constant time operation, compared to O(n) complexity with the elements of the contextVersionMap now.